mlflow/server/js/src/Actions.js

```diff
@@ -185,16 +185,36 @@ export const getUUID = () => {
  * Jquery's ajax promise is a bit weird so I chose to create a new Promise which resolves and
  * rejects using the ajax callbacks `success` and `error`.
  */
-const wrapDeferred = (deferred, data) => {
+export const wrapDeferred = (deferred, data, timeLeftMs = 60000, sleepMs = 1000) => {
   return new Promise((resolve, reject) => {
     deferred({
       data,
-      success: response => resolve(response),
-      error: xhr => {
+      success: response => {
+        resolve(response);
+      },
+      error: (xhr) => {
+        if (xhr.status === 429) {
+          if (timeLeftMs > 0) {
+            console.warn("Request failed with status code 429, message " +
+                new ErrorWrapper(xhr).getUserVisibleError() + ". Retrying after " +
+                sleepMs + " ms. On additional 429 errors, will continue to retry for up " +
+                "to " + timeLeftMs + " ms.");
+            // Retry the request, subtracting the current sleep duration from the remaining time
+            // and doubling the sleep duration
+            const newTimeLeft = timeLeftMs - sleepMs;
+            const newSleepMs = Math.min(newTimeLeft, sleepMs * 2);
+            return new Promise(resolveRetry => setTimeout(resolveRetry, sleepMs)).then(() => {
```

Comment:
I don't really know js that well. Why do you need the second Promise? Can you not just directly return the results of the recursive call?

Reply:
Ah the second promise (the `new Promise()` here) is effectively a means of implementing a `Thread.sleep` in Javascript - JS has no way to block the currently running thread, so we instead create a promise that resolves itself after a timeout using `setTimeout`, and then chain the recursive call to run afterwards. Note that `return`ing from the AJAX error callback on this line isn't actually necessary - the completion of the outer Promise happens when we call `resolve` or `reject` on lines 209 or 210.

If we didn't have the need to sleep, we'd still need to condition the resolution of our outer promise (i.e. the calls to `resolve` and `reject`) on the result of the recursive call, e.g. via

```
return wrapDeferred(deferred, data, newTimeLeft, newSleepMs).then(
                (successResponse) => resolve(successResponse),
                (failureResponse) => reject(failureResponse)
            );
```

Reply:
BTW, the explanation for how returning a `promise` from `then` works is in the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then#Return_value):

> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then#Return_value